crypto/engine/eng_openssl.c

```diff
@@ -513,22 +519,28 @@ static int ossl_hmac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
     dctx = EVP_PKEY_CTX_get_data(dst);
     dctx->md = sctx->md;
     if (!HMAC_CTX_copy(dctx->ctx, sctx->ctx))
-        return 0;
+        goto err;
     if (sctx->ktmp.data) {
         if (!ASN1_OCTET_STRING_set(&dctx->ktmp,
                                    sctx->ktmp.data, sctx->ktmp.length))
-            return 0;
+            goto err;
     }
     return 1;
+err:
+    ossl_hmac_cleanup(dst);
```

Comment:
If ossl_hmac_init(dst) is successful, it allocates memory for dst->data and a new ctx for the dst->data using HMAC_CTX_new().
If code reaches err:, ossl_hmac_init(dst) was successful, which would make EVP_PKEY_CTX_get_data(dst) a valid pointer which should be released (including the allocated HMAC_CTX).
Yes, in ossl_hmac_clean() sets data to NULL, ater EVP_PKEY_CTX_get_data(ctx) is released. This avoids the possibility of having a dangling pointer.